File: 05/zig/src/main.zig
const std = @import("std");

const Range = struct {
    low: i64,
    high: i64,
};

pub fn main() !void {
    var file = try std.fs.cwd().openFile("input-real.txt", .{});
    defer file.close();

    var buf: [1024]u8 = undefined;
    var file_reader = file.reader(&buf);

    // var result1: i32 = 0;
    // var result2: i64 = 0;

    // var ranges: []Range = &[_]Range{};
    var ranges = std.ArrayList(Range).initCapacity(std.heap.page_allocator, 16) catch unreachable;
    defer ranges.deinit(std.heap.page_allocator);

    while (try file_reader.interface.takeDelimiter('\n')) |line| {
        if (std.mem.eql(u8, line, "")) {
            break;
        }
        var iter = std.mem.splitScalar(u8, line, '-');
        const low = try std.fmt.parseInt(i64, iter.next().?, 10);
        const high = try std.fmt.parseInt(i64, iter.next().?, 10);
        try ranges.append(std.heap.page_allocator, Range{ .low = low, .high = high });
    }
    const rangeArray = ranges.items;

    std.mem.sort(Range, rangeArray, {}, cmpRange);

    const start = std.time.microTimestamp();
    const result2 = step_part2(rangeArray);
    const end = std.time.microTimestamp();

    std.debug.print("result2 {}, diff {}\n", .{result2, (end - start)});

    // std.debug.print("result1 {}\n", .{result1});
    // std.debug.print("result2 {}\n", .{result2});
}

pub fn cmpRange(_: void, a: Range, b: Range) bool {
    return (a.low < b.low);
}

pub fn step_part2(input: []Range) i64 {
    // const copy: []Range = &[_]Range{};
    // std.mem.copyForwards(Range, copy, input);
    var res: i64 = 0;
    var comparisons: i32 = 0;

    for (0..input.len) |index| {
        if (input[index].low == 0 and input[index].high == 0) {
            continue;
        }
        inner: for ((index+1)..input.len) |index2| {
            if (input[index2].low == 0 and input[index2].high == 0) {
                continue;
            }
            if (input[index].high < input[index2].low) {
                continue;
            }
            if (input[index].low > input[index2].high) {
                break :inner;
            }
            comparisons += 1;
            if (input[index].high >= input[index2].high and input[index].low <= input[index2].low) {
                input[index2].low = 0;
                input[index2].high = 0;
                continue;
            }
            if (input[index].low <= input[index2].low) {
                input[index2].low = input[index].high + 1;
                continue;
            }
            if (input[index].high >= input[index2].high) {
                input[index2].high = input[index].low - 1;
                continue;
            }
        }
    }

    for (input) |r| {
        if (r.low == 0 and r.high == 0) {
            continue;
        }
        res += r.high - r.low + 1;
    }
    std.debug.print("comparisons {}, size {}\n", .{comparisons, input.len});

    return res;
}

test "simple test part2" {
    // edge cases
    // try std.testing.expectEqual(92, try step_part2("911112", 2));

    // sample
}

File: 05/go/main.go
package main

import (
	"bufio"
	"fmt"
	"log/slog"
	"os"
	"slices"
	"strconv"
	"strings"
	"time"
)

func main() {
	err := run()
	if err != nil {
		slog.Error("failed", slog.String("error", err.Error()))
	}
}

type Range struct {
	Low  int
	High int
}

func run() error {
	f, err := os.Open("input-real.txt")
	if err != nil {
		return err
	}

	defer f.Close()

	scanner := bufio.NewScanner(f)

	result1 := 0
	ranges := []Range{}

	for scanner.Scan() {
		if strings.TrimSpace(scanner.Text()) == "" {
			break
		}
		line := scanner.Text()
		split := strings.Split(line, "-")
		l, err := strconv.ParseInt(split[0], 10, 64)
		if err != nil {
			return err
		}
		h, err := strconv.ParseInt(split[1], 10, 64)
		if err != nil {
			return err
		}
		ranges = append(ranges, Range{int(l), int(h)})
	}
	slices.SortStableFunc(ranges, func(a Range, b Range) int {
		if a.Low < b.Low {
			return -1
		}
		if a.Low > b.Low {
			return 1
		}
		return 0
	})
	// fmt.Println(ranges)

	for scanner.Scan() {
		// fmt.Println(scanner.Text())
		line := scanner.Text()
		num, err := strconv.ParseInt(line, 10, 64)
		if err != nil {
			return err
		}
		res := stepPart1(ranges, int(num))
		// fmt.Println(num, ",", res)

		if res {
			result1++
		}
	}
	fmt.Println("result1", result1)

	start := time.Now()
	res := stepPart2Alt(ranges)
	duration := time.Since(start)
	fmt.Println("result2", res, "time", duration)
	return nil
}

func stepPart1(ranges []Range, n int) bool {
	for _, r := range ranges {
		// fmt.Println("comparing", n, "with range", r)
		if n >= r.Low && n <= r.High {
			// fmt.Println("found", n)
			return true
		}
	}
	return false
}

func stepPart2Alt(ranges []Range) int {
	rangesCopy := slices.Clone(ranges)
	res := 0
	comparisons := 0
	for i, r1 := range rangesCopy {
		for j, r2 := range rangesCopy {
			if i == j {
				continue
			}
			if r2.High < r1.Low {
				continue
			}
			if r2.Low > r1.High {
				break
			}
			comparisons++
			if r1.High >= r2.High && r1.Low <= r2.Low {
				rangesCopy[j].Low = 0
				rangesCopy[j].High = 0
				continue
			}
			if r1.High >= r2.Low && r1.Low <= r2.Low {
				rangesCopy[j].Low = r1.High + 1
				continue
			}
			if r1.Low <= r2.High && r1.High >= r2.High {
				rangesCopy[j].High = r1.Low - 1
				continue
			}
		}
	}
	fmt.Println("comparisons", comparisons, "size", len(ranges))
	for _, r := range rangesCopy {
		if r.Low == 0 && r.High == 0 {
			continue
		}
		res += r.High - r.Low + 1
	}
	return res
}
